pipelines/salesforce/__init__.py

```diff
@@ -0,0 +1,226 @@
+                    ).strftime("%Y-%m-%dT%H:%M:%S.%fZ")
+        yield page
+        n_records += len(page)
+
+        # Update the state with the last replication key
+        if replication_key and page:
+            # Assume the records are sorted by replication key
+            state["last_replication_key"] = max(
+                state.get("last_replication_key", ""),
+                page[-1][replication_key],
+            )
+
+    print(f"Done with {sobject} extraction. {n_records} records extracted.")
+    if state and n_records:
+        print("New state:", state)
+
+
+def sobject_factory(
+    name: str,
+    sobject: str,
+    client: Salesforce,
+    write_disposition: str = "replace",
+    replication_key: t.Optional[str] = None,
+    deferrable: bool = False,
+    fields: t.Optional[t.List[str]] = None,
+) -> t.Callable[[], t.Iterable[dict]]:
+    """Return a resource for a Salesforce sobject.
+
+    Args:
+        name: The name of the resource. This affects the table name in the warehouse.
+        sobject: The Salesforce object API name.
+        client: A Salesforce client.
+        write_disposition: The write disposition for the resource.
+        replication_key: The field to use for incremental replication.
+        deferrable: Whether the resource should be deferrable. We recommend not
+            deferring large resources as it could cause memory issues.
+        fields: A list of fields to extract. If not specified, all fields
+            will be extracted.
+    """
+
+    @dlt.resource(name=name, write_disposition=write_disposition)
+    def sobject_resource():
+        nonlocal write_disposition, replication_key
+
+        if replication_key and write_disposition == "append":
+            state: dict = dlt.state().setdefault(name, {})
+        else:
+            replication_key = state = None
+
+        if deferrable:
+            yield dlt.defer(
+                lambda: [
+                    record
+                    for page in get_records(
+                        client, sobject, state, replication_key, fields
+                    )
+                    for record in page
+                ]
+            )
+        else:
+            yield from get_records(client, sobject, state, replication_key, fields)
+
+    return sobject_resource
+
+
+@dlt.source(name="salesforce")
+def salesforce_source(
+    username: str = dlt.secrets.value,
+    password: str = dlt.secrets.value,
+    security_token: str = dlt.secrets.value,
+):
+    client = Salesforce(username, password, security_token)
+    # You can add more sobjects here, custom ones too
+    # See standard objects here:
+    # https://developer.salesforce.com/docs/atlas.en-us.object_reference.meta/object_reference/sforce_api_objects_list.htm
+    # See the `sobject_factory` function docstring for info on the arguments
+    return (
+        sobject_factory(name="sf_user", sobject="User", client=client, deferrable=True),
```

Comment:
the table name now is sf_user, because I could not get the table count query to work with all destinations at the same time. for duckdb a table named user must be escaped, for bigquery the escapes do not work, but an unescaped table named user does work.

Reply:
BigQuery uses different quotation (``) to escape identifiers. we can use have `user` if you use `sql_client.make_qualified_table_name` to escape the table name.

yeah but let's keep sf_user. it will make writing queries easier